spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MultipleInstantiationsOfSingletons.java

```diff
@@ -0,0 +1,228 @@
+    private List<XMethod> methodsUsingMonitor;
+    private Set<XMethod> constructors;
+
+    public MultipleInstantiationsOfSingletons(BugReporter bugReporter) {
+        this.bugReporter = bugReporter;
+
+        try {
+            cloneableInterface = Repository.getInterfaces("java.lang.Cloneable")[0];
+            serializableInterface = Repository.getInterfaces("java.io.Serializable")[0];
+        } catch (ClassNotFoundException e) {
+            bugReporter.reportMissingClass(e);
+        }
+    }
+
+    @Override
+    public void visit(JavaClass obj) {
+        isSingleton = false;
+        isGetterMethodSynchronized = false;
+
+        isCloneable = false;
+        implementsCloneableDirectly = false;
+        hasCloneMethod = false;
+        cloneOnlyThrowsException = false;
+        cloneOnlyThrowsCloneNotSupportedException = false;
+
+        isSerializable = false;
+
+        constructors = new HashSet<>();
+        methods = new EnumMap<>(Methods.class);
+        methodsUsingMonitor = new ArrayList<>();
+
+        if (obj.getClassName().endsWith("Singleton")) {
+            isSingleton = true;
+        }
+
+        // Does this class directly implement Cloneable or Serializable?
+        try {
+            JavaClass[] interfaces = obj.getAllInterfaces();
+            isCloneable = java.util.stream.Stream.of(interfaces).anyMatch(i -> i.equals(cloneableInterface));
+            isSerializable = java.util.stream.Stream.of(interfaces).anyMatch(i -> i.equals(serializableInterface));
+
+            implementsCloneableDirectly = java.util.stream.Stream.of(obj.getInterfaces()).anyMatch(i -> i.equals(cloneableInterface));
+        } catch (ClassNotFoundException e) {
+            bugReporter.reportMissingClass(e);
+        }
+
+        super.visit(obj);
+    }
+
+    @Override
+    public void visit(Method obj) {
+        if ("clone".equals(getMethodName()) && "()Ljava/lang/Object;".equals(getMethodSig())) {
+            cloneOnlyThrowsException = PruneUnconditionalExceptionThrowerEdges.doesMethodUnconditionallyThrowException(getXMethod());
+            methods.put(Methods.CLONE, getXMethod());
+            hasCloneMethod = true;
+        }
+
+        if (Const.CONSTRUCTOR_NAME.equals(getMethodName())) {
+            constructors.add(getXMethod());
+        }
+
+        super.visit(obj);
+    }
+
+    @Override
+    public boolean beforeOpcode(int seen) {
+        if (seen == Const.ATHROW && "clone".equals(getMethodName())) {
+            return true;
+        }
+        if (seen == Const.ARETURN || seen == Const.MONITORENTER) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void sawOpcode(int seen) {
+        if (seen == Const.ATHROW && stack.getStackDepth() > 0) {
+            OpcodeStack.Item item = stack.getStackItem(0);
+            if (item != null && item.getSignature().equals("Ljava/lang/CloneNotSupportedException;") && cloneOnlyThrowsException) {
```

Comment:
It's a good practice to compare to the constant value to avoid NPE, not the other way around. (Just like you did with `clone` in line 116 and 131.)